belief_manager.py

```diff
@@ -0,0 +1,578 @@
+
+    self._fetch_autopilot_data()
+    # Initialize the Kalman filter.
+    self._filter = Kalman((self._observed_position_x,
+                            self._observed_position_y),
+                           (self._observed_velocity_x,
+                            self._observed_velocity_y))
+
+  def _initialize_member_variables(self):
+    """ Initializes essential member variables. This is mostly so that we can
+    call this function during testing without calling all of __init__. """
+    # Keeps track of past good LOB readings. The list lines up 1-to-1 with the
+    # order of LOBs in the state. Each item is a tuple containing the
+    # LOB reading, its variance, and the x and y components of the drone's
+    # current position when this LOB was accurate. These will never be the
+    # current state.
+    self._past_lobs = []
+    # A record of the state we last used to update _past_lobs.
+    self._old_state = None
+    # Keeps a record of past states.
+    self._past_states = collections.deque()
+    # A record of the number of transmitters we've seen for the past few
+    # cycles.
+    self._observed_transmitters = collections.deque()
+    # Error regions calculated from our last cycle.
+    self._last_error_regions = {}
+    # A dictionary of transmitter indices. Each value is the cycle we last got
+    # data for it.
+    self._cycle_data = {}
+
+    # How many cycles we've run.
+    self._cycles = 0
+
+  def _fetch_autopilot_data(self):
+    """ Fetches measurements from the onboard Pixhawk autopilot, and updates the
+    member variables for vehicle position and velocity observations accordingly.
+    """
+    # Read the latest data from the autopilot.
+    sensor_data = self.__autopilot.read_front()
+
+    self._observed_position_x = sensor_data.latitude
+    self._observed_position_y = sensor_data.longitude
+
+    self._observed_velocity_x = sensor_data.velocity[self._X]
+    self._observed_velocity_y = sensor_data.velocity[self._Y]
+
+    logger.info("Got new sensor data: position: (%f, %f), velocity: (%f, %f)" \
+                % (self._observed_position_x, self._observed_position_y,
+                   self._observed_velocity_x, self._observed_velocity_y))
+
+  def _fetch_radio_data(self):
+    """ Fetches measurements from the radio.
+    Returns:
+        A list containing one tuple for every object sighted by the radar.
+        Each tuple contains the LOB measurement and signal strength. """
+    transmitters = []
+
+    # Keep reading data until we have no more to read.
+    while True:
+      signal = self.__radio.read_next(block=False)
+      if not signal:
+        # No more data.
+        break
+
+      # Normalize the LOB, so that the angle will still be correct even if the
+      # drone starts flying in a different direction.
+      nomalized_lob = self._filter.normalize_lobs(signal.lob)
+
+      transmitters.append((normalized_lob, signal.strength))
```

Comment:
1. The _associate_lobs method is designed to handle this case.
2. For now, we're assuming that velocity remains constant, i.e. the drone is always pointing in the same direction. (Handling changing velocities will be coming at a future time...) I'll put a TODO in there reminding me to change it when the time comes, though.